core/src/main/java/io/opencensus/tags/TagContext.java

```diff
@@ -135,7 +145,7 @@ public Builder clear(TagKey<?> key) {
      * @return a {@code TagContext} with the same tags as this builder.
      */
     public TagContext build() {
-      return new TagContext(new HashMap<TagKey<?>, Object>(tags));
+      return new TagContext(new HashMap<TagKey, Object>(tags));
```

Comment:
You copy the tags at least one extra time here + in TagContext ctor.

Also another optimization may be to never copy the tags map in the TagContext and make builder be invalid after called build (don't know if that worths the complication).